### ISSUE TITLE
A note about not firing side effects in useState(prev => next) ?

### ISSUE THREAD

**Issue by dxinteractive** [reactjs/reactjs.org#2882]
So I ended up doing this by mistake after a wonky merge:

```
let Foo = (props) => {
    let [bar, setBar] = useState();
    
    let handleClick = () => {
        useState(bar => {
            let newBar = bar + 1;
            props.handleBarChange(newBar); // bad, which isn't surprising
            return newBar;
        });
    };

    return <div onClick={handleClick}>click me</div>;
};
```

If you fire a side effect such as calling something that sets state in a higher component from a useState updater function, you get the error "Cannot update a component from inside the function body of a different component" (but not always! It seems to depend on internal timing). This definitely seems like a bad idea when you look at it, but I couldn't find any mention in the docs that this is bad.

https://reactjs.org/docs/hooks-reference.html#functional-updates

I initially wanted to comment about this here (https://github.com/facebook/react/issues/18178), because the general message of that thread was "don't set state of another component in a render method", but I've managed to get this error message without doing that. Thanks.

**Comment by gaearon** [comment]
Updaters generally run in the render phase, while a component is rendering. So doing any side effects there or setting state on other components is definitely a no-no.

**Comment by dxinteractive** [comment]
Makes sense. I guess the missing piece of info for me was whether a setState call was regarded as occurring in the render phase, or merely _triggering_ a render phase and therefore wasn't part of the render phase itself. Knowing that it's easy to prove that the code up there is a silly thing to do.

While the terms "render phase" and "commit phase" are mentioned in the docs a couple of places, I wasn't able to find anything that clearly said "all these things are in the render phase / all these other things are in the commit phase / and these things cause transitions between phases". The React Component page alludes to it, but only in passing. I did find this diagram which if it's correct seems quite useful http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/

I'd love to see a page in the docs that walks people through that slightly deeper knowledge of how the lifecycle works, and which lifecycle methods belong where, and which trigger other states. I really like how a bunch of the hooks documentation didn't sugar-coat how they worked, they gave a decent description of how hooks work internally so that devs get the right mental model from the start. And honestly React's docs have done a good job of that on a  conceptual level with the whole props-go-down, callbacks-go-up thing too.

Feel free to close if you feel like the docs are good enough though, or if the pay off for the time invested in a lifecycle overview / diagram page would not we worth it for how many people would benefit from it. I could very well be in the minority where it comes to wanting to knowing what things truly count as the render phase vs the commit phase.
